superset/common/query_object.py

```diff
@@ -0,0 +1,59 @@
+# pylint: disable=R
+from typing import Dict, List, Optional, Union
+
+from superset import app
+from superset.utils import core as utils
+
+# TODO: Type Metrics dictionary with TypedDict when it becomes a vanilla python type
+# https://github.com/python/mypy/issues/5288
+Metric = Union[str, Dict]
+
+
+class QueryObject:
+    """
+    The query object's schema matches the interfaces of DB connectors like sqla
+    and druid. The query objects are constructed on the client.
+    """
+    def __init__(
+            self,
+            granularity: str,
+            groupby: List[str],
+            metrics: List[Metric],
+            filters: List[str],
+            time_range: Optional[str] = None,
```

Comment:
do we need to support `time_range` _and_ `since` + `until`? in an ideal world it'd be nice to not have time ranges coming from two different places.

Reply:
Good question! 

I agree we should only have one. Looking at the code, seems like the calculation based on `time_range` encompass the one based on `since` and `until`. I'll have the new `query_object` only take `time_range`. 😄

Reply:
sweet!